File: app/src/main/java/com/faendir/lightning_launcher/multitool/animation/Transformation.java
package com.faendir.lightning_launcher.multitool.animation;

import android.graphics.PointF;
import android.view.View;
import android.view.ViewPropertyAnimator;
import com.faendir.lightning_launcher.multitool.proxy.Item;
import com.faendir.lightning_launcher.multitool.proxy.PropertySet;

/**
 * @author lukas
 * @since 10.07.18
 */
class Transformation {
    PointF scale = new PointF(1, 1);
    PointF translate = new PointF(0, 0);
    PointF pivot = new PointF(0, 0);
    float turn = 0;
    PointF rotation = new PointF(0, 0);
    float alpha = 1;
    private boolean partial = true;

    void transform(Item item) {
        String pinMode = item.getProperties().getString(PropertySet.ITEM_PIN_MODE);
        PointB transform = new PointB(!pinMode.contains("X"), !pinMode.contains("Y"));
        if (partial ? transform.any() : transform.both()) {
            PointF center = AnimationScript.center(item, false);
            transform(item.getRootView(), transform, center);
        }
    }

    void transform(View view, PointB transform, PointF center) {
        view.setPivotX(center.x + pivot.x);
        view.setPivotY(center.y + pivot.y);
        ViewPropertyAnimator animator = view.animate().setDuration(0).alpha(alpha).rotation(turn * 360);
        if (transform.x) animator.scaleX(scale.x).translationX(translate.x).rotationX(rotation.x * 360);
        if (transform.y) animator.scaleY(scale.y).translationY(translate.y).rotationY(rotation.y * 360);
        animator.start();
    }

    Transformation onlyUnpinnedItems() {
        partial = false;
        return this;
    }
}

File: app/src/main/java/com/faendir/lightning_launcher/multitool/animation/Transformation.kt
package com.faendir.lightning_launcher.multitool.animation

import android.graphics.PointF
import android.view.View
import com.faendir.lightning_launcher.multitool.proxy.Item
import com.faendir.lightning_launcher.multitool.proxy.PropertySet

/**
 * @author lukas
 * @since 10.07.18
 */
class Transformation {
    var scale = PointF(1f, 1f)
    var translate = PointF(0f, 0f)
    var pivot = PointF(0f, 0f)
    var turn = 0f
    var rotation = PointF(0f, 0f)
    var alpha = 1f
    private var partial = true

    fun transform(item: Item) {
        val pinMode = item.properties.getString(PropertySet.ITEM_PIN_MODE)
        val transform = PointB(!pinMode.contains("X"), !pinMode.contains("Y"))
        if (if (partial) transform.any() else transform.both()) {
            val center = AnimationScript.center(item, false)
            transform(item.rootView, transform, center)
        }
    }

    fun transform(view: View, transform: PointB, center: PointF) {
        view.pivotX = center.x + pivot.x
        view.pivotY = center.y + pivot.y
        val animator = view.animate().setDuration(0).alpha(alpha).rotation(turn * 360)
        if (transform.x) animator.scaleX(scale.x).translationX(translate.x).rotationX(rotation.x * 360)
        if (transform.y) animator.scaleY(scale.y).translationY(translate.y).rotationY(rotation.y * 360)
        animator.start()
    }

    fun onlyUnpinnedItems(): Transformation {
        partial = false
        return this
    }
}
